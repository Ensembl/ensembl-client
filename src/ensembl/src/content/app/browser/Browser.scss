@import 'src/styles/common';

.browser {
  width: 100%;
  height: calc(100% - 32px);
  position: relative;
}

.browserInnerWrapper {
  position: absolute;
  overflow: hidden;
  width: 100vw;
  @include flex;
  transition: height 0.3s linear;
  &.shorter {
    height: calc(100vh - 194px);
  }

  &.taller {
    height: calc(100vh - 146px);
  }
}

.browserImageWrapper {
  height: 100%;
}

.browserOverlay {
  background: $ens-white;
  height: 100%;
  opacity: 0.7;
  position: absolute;
  width: 100vw;
  z-index: 150;
}

.exampleLinks {
  margin-top: 50px;
  margin-left: 40px;
  dd {
    margin-bottom: 30px;
  }
}
.objectType {
  color: $ens-grey;
  margin-right: 15px;
  width: 60px;
  text-align: right;
  display: inline-block;
}

.objectLabel {
  color: $ens-blue;
}
