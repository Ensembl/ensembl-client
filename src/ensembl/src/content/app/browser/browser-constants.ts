export const BROWSER_CONTAINER_ID = 'genome-browser';
