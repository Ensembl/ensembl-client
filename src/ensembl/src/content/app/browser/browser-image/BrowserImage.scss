.browserImagePlus {
  height: 100%;
  position: relative;
}

.browserStage {
  height: 100%;
  width: 100%;
  overflow: hidden;
  z-index: -1;

  &.shorter {
    height: calc(100% - 56px);
  }
}

.browserInnerWrapper {
  overflow: hidden;
}

.loaderWrapper {
  left: calc(50% - 20px);
  top: calc(50% - 20px);
  position: absolute;
}
