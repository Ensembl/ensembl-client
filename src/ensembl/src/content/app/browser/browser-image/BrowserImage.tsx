import React, { PureComponent } from 'react';

import styles from './BrowserImage.scss';

type BrowserImageProps = {};

class BrowserImage extends PureComponent<BrowserImageProps> {
  private browserCanvas: React.RefObject<HTMLDivElement>;

  constructor(props: BrowserImageProps) {
    super(props);

    this.browserCanvas = React.createRef();
  }

  public componentDidMount() {
    const currentEl = this.browserCanvas.current;
    this.activateIfPossible(currentEl);
  }

  public render() {
    return <div className={styles.browserStage} ref={this.browserCanvas} />;
  }

  private activateIfPossible(currentEl: HTMLElement | null) {
    const activateEvent = new CustomEvent('bpane-activate', {
      bubbles: true,
      detail: {
        key: 'main'
      }
    });

    let done = false;

    if (currentEl && currentEl.ownerDocument) {
      const browserEl = currentEl.ownerDocument.querySelector(
        '.browser-stage'
      ) as HTMLBodyElement;

      const bodyEl = currentEl.ownerDocument.querySelector('body');

      if (bodyEl && bodyEl.classList.contains('browser-app-ready')) {
        browserEl.dispatchEvent(activateEvent);
        done = true;
      }
    }

    if (!done) {
      setTimeout(() => this.activateIfPossible(currentEl), 50);
    }
  }
}

export default BrowserImage;
