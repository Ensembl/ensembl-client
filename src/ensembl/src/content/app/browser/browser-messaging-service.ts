import windowService, {
  WindowServiceInterface
} from 'src/services/window-service';

import JSONValue from 'src/shared/types/JSON';

/*
  This is a service for communicating between genome browser and React wrapper.
*/

export class BrowserMessagingService {
  private window: Window;
  private isRecepientReady: boolean = false;
  private subscribers: any = {};
  private outgoingMessageQueue: JSONValue[] = [];

  public constructor(windowService: WindowServiceInterface) {
    this.window = windowService.getWindow();
    this.subscribeToMessages();
    this.subscribe('bpane-ready', this.onRecepientReady);
    this.ping();
  }

  private ping() {
    this.sendPostMessage({ type: 'bpane-ready-query' });
  }

  private onRecepientReady = () => {
    console.log('RECEIVED PONG!!!');
    this.isRecepientReady = true;

    this.outgoingMessageQueue.forEach((message) =>
      this.sendPostMessage(message)
    );
  };

  private sendPostMessage(message: JSONValue) {
    console.log('SENDING MESSAGE', message);
    this.window.postMessage(message, '*');
  }

  private subscribeToMessages() {
    this.window.addEventListener('message', this.handleMessage);
  }

  private addMessageToQueue(message: JSONValue) {
    this.outgoingMessageQueue.push(message);
  }

  private handleMessage = (event: MessageEvent) => {
    const {
      data: { type, payload }
    } = event;
    if (!(type && payload)) {
      return;
    }
    const subscribers = this.subscribers[type];
    if (subscribers) {
      subscribers.forEach((subscriber: Function) => subscriber(payload));
    }
  };

  public subscribe = (eventName: string, callback: Function) => {
    if (!this.subscribers[eventName]) {
      this.subscribers[eventName] = new Set();
      this.subscribers[eventName].add(callback);
    }

    return {
      unsubscribe: () => {
        this.subscribers[eventName].delete(callback);
      }
    };
  };

  public send = (eventName: string, payload: JSONValue) => {
    const message = {
      type: eventName,
      payload
    };

    if (!this.isRecepientReady) {
      this.addMessageToQueue(message);
    } else {
      this.sendPostMessage(message);
    }
  };
}

export default new BrowserMessagingService(windowService);
