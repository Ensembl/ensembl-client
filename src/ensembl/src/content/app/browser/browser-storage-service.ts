import storageService, {
  StorageServiceInterface
} from 'src/services/storage-service';
import {
  TrackStates,
  TrackType,
  TrackToggleStates
} from './track-panel/trackPanelConfig';
import { ChrLocation } from './browserState';
import { ImageButtonStatus } from 'src/shared/image-button/ImageButton';

export enum StorageKeys {
  ACTIVE_GENOME_ID = 'browser.activeGenomeId',
  ACTIVE_ENS_OBJECT_ID = 'browser.activeEnsObjectId',
  CHR_LOCATION = 'browser.chrLocation',
  DEFAULT_CHR_LOCATION = 'browser.defaultChrLocation',
  TRACK_STATES = 'browser.trackStates',
  TRACK_LIST_TOGGLE_STATES = 'browser.trackListToggleStates',
  SELECTED_BROWSER_TAB = 'browser.selectedBrowserTab'
}

export class BrowserStorageService {
  private storageService: StorageServiceInterface;

  public constructor(storageService: StorageServiceInterface) {
    this.storageService = storageService;
  }

  public getActiveGenomeId(): string {
    return this.storageService.get(StorageKeys.ACTIVE_GENOME_ID);
  }

  public saveActiveGenomeId(activeGenomeId: string) {
    this.storageService.save(StorageKeys.ACTIVE_GENOME_ID, activeGenomeId);
  }

  public getActiveEnsObjectId() {
    return this.storageService.get(StorageKeys.ACTIVE_ENS_OBJECT_ID) || {};
  }

  public updateActiveEnsObjectId(activeEnsObjectId: {
    [genomeId: string]: string;
  }) {
    this.storageService.update(
      StorageKeys.ACTIVE_ENS_OBJECT_ID,
      activeEnsObjectId
    );
  }

  public getChrLocation() {
    return this.storageService.get(StorageKeys.CHR_LOCATION) || {};
  }

  public updateChrLocation(chrLocation: { [genomeId: string]: ChrLocation }) {
    this.storageService.update(StorageKeys.CHR_LOCATION, chrLocation);
  }

  public getDefaultChrLocation() {
    return this.storageService.get(StorageKeys.DEFAULT_CHR_LOCATION) || {};
  }

  public updateDefaultChrLocation(defaultChrLocation: {
    [genomeId: string]: ChrLocation;
  }) {
    this.storageService.update(
      StorageKeys.DEFAULT_CHR_LOCATION,
      defaultChrLocation
    );
  }

  public getTrackStates(): TrackStates {
    return this.storageService.get(StorageKeys.TRACK_STATES) || {};
  }

  public saveTrackStates(
    genomeId: string,
    categoryName: string,
    trackName: string,
    trackStatus: ImageButtonStatus
  ) {
    const trackStates = this.getTrackStates();

    if (!trackStates[genomeId]) {
      trackStates[genomeId] = {};
    }

    if (!trackStates[genomeId][categoryName]) {
      trackStates[genomeId][categoryName] = {};
    }

    trackStates[genomeId][categoryName][trackName] = trackStatus;

    this.storageService.save(StorageKeys.TRACK_STATES, trackStates);
  }

  public getTrackListToggleStates() {
    return this.storageService.get(StorageKeys.TRACK_LIST_TOGGLE_STATES) || {};
  }

  public updateTrackListToggleStates(toggleState: {
    [genomeId: string]: TrackToggleStates;
  }) {
    this.storageService.update(
      StorageKeys.TRACK_LIST_TOGGLE_STATES,
      toggleState
    );
  }

  public getSelectedBrowserTab() {
    return (
      this.storageService.get(StorageKeys.SELECTED_BROWSER_TAB) ||
      TrackType.GENOMIC
    );
  }

  public updateSelectedBrowserTab(selectedBrowserTabForGenome: {
    [genomeId: string]: TrackType;
  }) {
    this.storageService.update(
      StorageKeys.SELECTED_BROWSER_TAB,
      selectedBrowserTabForGenome
    );
  }
}

export default new BrowserStorageService(storageService);
