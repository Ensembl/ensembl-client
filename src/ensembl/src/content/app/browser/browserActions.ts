import { createAction } from 'typesafe-actions';
import { Dispatch, ActionCreator, Action } from 'redux';
import { ThunkAction } from 'redux-thunk';

import config from 'config';

import { BrowserNavStates, ChrLocation, CogList } from './browserState';
import {
  getBrowserActiveGenomeId,
  getBrowserActiveEnsObjectId,
  getDefaultChrLocation,
  getChrLocation
} from './browserSelectors';
import { getBrowserAnalyticsObject } from 'src/analyticsHelper';
import browserStorageService from './browser-storage-service';
import { RootState } from 'src/store';

export const updateBrowserActivated = createAction(
  'browser/update-browser-activated',
  (resolve) => {
    return (browserActivated: boolean) =>
      resolve(browserActivated, getBrowserAnalyticsObject('Default Action'));
  }
);

export const activateBrowser = (browserEl: HTMLDivElement) => {
  return (dispatch: Dispatch) => {
    const { protocol, host: currentHost } = location;
    const host = config.apiHost || currentHost;
    const activateEvent = new CustomEvent('bpane-activate', {
      bubbles: true,
      detail: {
        'config-url': `${protocol}${host}/browser/config`,
        key: 'main'
      }
    });

    browserEl.dispatchEvent(activateEvent);

    dispatch(updateBrowserActivated(true));
  };
};

export const updateBrowserActiveGenomeId = createAction(
  'browser/update-active-genome-id',
  (resolve) => {
    return (activeGenomeId: string) =>
      resolve(activeGenomeId, getBrowserAnalyticsObject('Navigation'));
  }
);

export const updateBrowserActiveGenomeIdAndSave: ActionCreator<
  ThunkAction<void, any, null, Action<string>>
> = (activeGenomeId: string) => (dispatch: Dispatch) => {
  dispatch(updateBrowserActiveGenomeId(activeGenomeId));
  browserStorageService.saveActiveGenomeId(activeGenomeId);
};

export const updateBrowserActiveEnsObjectId = createAction(
  'browser/update-active-ens-object-id',
  (resolve) => {
    return (activeEnsObjectId: { [objectId: string]: string }) =>
      resolve(activeEnsObjectId, getBrowserAnalyticsObject('Navigation'));
  }
);

export const updateBrowserActiveEnsObjectIdAndSave: ActionCreator<
  ThunkAction<void, any, null, Action<string>>
> = (activeEnsObjectId: string) => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const currentActiveEnsObjectId = getBrowserActiveEnsObjectId(getState());
    const updatedActiveEnsObjectId = { ...currentActiveEnsObjectId };
    const genomeId = activeEnsObjectId.split(':')[0];
    updatedActiveEnsObjectId[genomeId] = activeEnsObjectId;

    dispatch(updateBrowserActiveEnsObjectId(updatedActiveEnsObjectId));

    browserStorageService.updateActiveEnsObjectId(updatedActiveEnsObjectId);
  };
};

export const toggleBrowserNav = createAction(
  'browser/toggle-browser-navigation',
  (resolve) => {
    return () => resolve(undefined, getBrowserAnalyticsObject('Navigation'));
  }
);

export const updateBrowserNavStates = createAction(
  'browser/update-browser-nav-states',
  (resolve) => {
    return (browserNavStates: BrowserNavStates) =>
      resolve(browserNavStates, getBrowserAnalyticsObject('Navigation'));
  }
);

export const updateChrLocation = createAction(
  'browser/update-chromosome-location',
  (resolve) => {
    return (chrLocationData: { [genomeId: string]: ChrLocation }) =>
      resolve(chrLocationData);
  }
);

export const setChrLocation: ActionCreator<
  ThunkAction<any, any, null, Action<string>>
> = (chrLocation: ChrLocation) => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const genomeId = getBrowserActiveGenomeId(getState());
    const currentChrLocation = getChrLocation(getState());
    const updatedChrLocation = { ...currentChrLocation };
    updatedChrLocation[genomeId] = chrLocation;

    dispatch(updateChrLocation(updatedChrLocation));
    browserStorageService.updateChrLocation(updatedChrLocation);
  };
};

export const updateDefaultChrLocation = createAction(
  'browser/update-default-chromosome-location',
  (resolve) => {
    return (chrLocationData: { [genomeId: string]: ChrLocation }) =>
      resolve(chrLocationData, getBrowserAnalyticsObject('User Interaction'));
  }
);

export const changeBrowserLocation: ActionCreator<
  ThunkAction<any, any, null, Action<string>>
> = (chrLocation: ChrLocation, browserEl: HTMLDivElement) => {
  return (dispatch: Dispatch, getState: () => RootState) => {
    const [chrCode, startBp, endBp] = chrLocation;
    const genomeId = getBrowserActiveGenomeId(getState());

    const stickEvent = new CustomEvent('bpane', {
      bubbles: true,
      detail: {
        stick: `${genomeId}:${chrCode}`
      }
    });

    browserEl.dispatchEvent(stickEvent);

    if (startBp > 0 && endBp > 0) {
      const gotoEvent = new CustomEvent('bpane', {
        bubbles: true,
        detail: {
          goto: `${startBp}-${endBp}`
        }
      });

      browserEl.dispatchEvent(gotoEvent);
    }

    const currentChrLocation = getChrLocation(getState());
    const updatedChrLocation = { ...currentChrLocation };
    updatedChrLocation[genomeId] = chrLocation;

    dispatch(updateChrLocation(updatedChrLocation));
    browserStorageService.updateChrLocation(updatedChrLocation);

    const currentDefaultChrLocation = getDefaultChrLocation(getState());

    if (!currentDefaultChrLocation[genomeId]) {
      const updatedDefaultChrLocation = { ...currentDefaultChrLocation };
      updatedDefaultChrLocation[genomeId] = chrLocation;
      dispatch(updateDefaultChrLocation(updatedDefaultChrLocation));
      browserStorageService.updateDefaultChrLocation(updatedDefaultChrLocation);
    }
  };
};

export const updateCogList = createAction(
  'browser/update-cog-list',
  (resolve) => {
    return (cogList: number) => {
      return resolve(cogList, getBrowserAnalyticsObject('User Interaction'));
    };
  }
);

export const updateCogTrackList = createAction(
  'browser/update-cog-track-list',
  (resolve) => {
    return (trackY: CogList) => {
      return resolve(trackY, getBrowserAnalyticsObject('User Interaction'));
    };
  }
);

export const updateSelectedCog = createAction(
  'browser/update-selected-cog',
  (resolve) => {
    return (index: string) => {
      return resolve(index, getBrowserAnalyticsObject('User Interaction'));
    };
  }
);

export const updateTrackConfigNames = createAction(
  'browser/update-track-config-names',
  (resolve) => {
    return (selectedCog: any, sense: boolean) => {
      return resolve(
        [selectedCog, sense],
        getBrowserAnalyticsObject('User Interaction')
      );
    };
  }
);

export const updateTrackConfigLabel = createAction(
  'browser/update-track-config-label',
  (resolve) => {
    return (selectedCog: any, sense: boolean) => {
      return resolve(
        [selectedCog, sense],
        getBrowserAnalyticsObject('User Interaction')
      );
    };
  }
);

export const updateApplyToAll = createAction(
  'browser/update-apply-to-all',
  (resolve) => {
    return (yn: boolean) => {
      return resolve(yn, getBrowserAnalyticsObject('User Interaction'));
    };
  }
);

export const toggleGenomeSelector = createAction(
  'toggle-genome-selector',
  (resolve) => {
    return (genomeSelectorActive: boolean) =>
      resolve(
        genomeSelectorActive,
        getBrowserAnalyticsObject('User Interaction')
      );
  }
);
