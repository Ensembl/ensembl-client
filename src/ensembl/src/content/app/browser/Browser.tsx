import React, { FunctionComponent, useRef, useEffect, useState } from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { connect } from 'react-redux';
import { replace, Replace } from 'connected-react-router';
import { useSpring, animated } from 'react-spring';
import { Link } from 'react-router-dom';
import BrowserBar from './browser-bar/BrowserBar';
import BrowserImage from './browser-image/BrowserImage';
import BrowserNavBar from './browser-nav/BrowserNavBar';
import TrackPanel from './track-panel/TrackPanel';
import AppBar from 'src/shared/app-bar/AppBar';
import find from 'lodash/find';

import { RootState } from 'src/store';
import {
  BrowserOpenState,
  BrowserNavStates,
  ChrLocation
} from './browserState';
import {
  changeBrowserLocation,
  updateBrowserNavStates,
  updateBrowserActiveGenomeIdAndSave,
  updateBrowserActiveEnsObjectIdAndSave
} from './browserActions';
import {
  getBrowserOpenState,
  getBrowserNavOpened,
  getChrLocation,
  getGenomeSelectorActive,
  getBrowserActivated,
  getBrowserActiveGenomeId,
  getBrowserQueryParams,
  getBrowserActiveEnsObjectId
} from './browserSelectors';
import { getLaunchbarExpanded } from 'src/header/headerSelectors';
import { getTrackPanelOpened } from './track-panel/trackPanelSelectors';
import { getChrLocationFromStr, getChrLocationStr } from './browserHelper';
import { getDrawerOpened } from './drawer/drawerSelectors';
import { getEnabledCommittedSpecies } from 'src/content/app/species-selector/state/speciesSelectorSelectors';
import { CommittedItem } from 'src/content/app/species-selector/types/species-search';
import {
  fetchEnsObject,
  fetchEnsObjectTracks,
  fetchExampleEnsObjects
} from 'src/ens-object/ensObjectActions';
import { getExampleEnsObjects } from 'src/ens-object/ensObjectSelectors';
import {
  ExampleEnsObjectsData,
  EnsObject
} from 'src/ens-object/ensObjectTypes';

import { getGenomeInfo } from 'src/genome/genomeSelectors';
import { GenomeInfoData } from 'src/genome/genomeTypes';
import {
  fetchGenomeInfo,
  fetchGenomeTrackCategories
} from 'src/genome/genomeActions';
import { toggleDrawer } from './drawer/drawerActions';

import browserStorageService from './browser-storage-service';
import { TrackStates } from './track-panel/trackPanelConfig';
import { AppName } from 'src/global/globalConfig';
import upperFirst from 'lodash/upperFirst';

import * as urlFor from 'src/shared/helpers/urlHelper';

import styles from './Browser.scss';

import 'static/browser/browser.js';

type StateProps = {
  activeGenomeId: string;
  activeEnsObjectId: { [genomeId: string]: string };
  browserActivated: boolean;
  browserNavOpened: boolean;
  browserOpenState: BrowserOpenState;
  browserQueryParams: { [key: string]: string };
  chrLocation: { [genomeId: string]: ChrLocation };
  drawerOpened: boolean;
  genomeInfo: GenomeInfoData;
  genomeSelectorActive: boolean;
  trackPanelOpened: boolean;
  launchbarExpanded: boolean;
  exampleEnsObjects: ExampleEnsObjectsData;
  committedSpecies: CommittedItem[];
};

type DispatchProps = {
  changeBrowserLocation: (
    chrLocation: ChrLocation,
    browserEl: HTMLDivElement
  ) => void;
  fetchEnsObject: (ensObjectId: string, genomeId: string) => void;
  fetchEnsObjectTracks: (ensObjectId: string, genomeId: string) => void;
  fetchExampleEnsObjects: (genomeId: string) => void;
  fetchGenomeInfo: (genomeId: string) => void;
  fetchGenomeTrackCategories: (genomeId: string) => void;
  replace: Replace;
  toggleDrawer: (drawerOpened: boolean) => void;
  updateBrowserActiveGenomeIdAndSave: (genomeId: string) => void;
  updateBrowserActiveEnsObjectIdAndSave: (objectId: string) => void;
  updateBrowserNavStates: (browserNavStates: BrowserNavStates) => void;
};

type OwnProps = {};

type MatchParams = {
  genomeId: string;
};

type BrowserProps = RouteComponentProps<MatchParams> &
  StateProps &
  DispatchProps &
  OwnProps;

export const Browser: FunctionComponent<BrowserProps> = (
  props: BrowserProps
) => {
  const browserRef: React.RefObject<HTMLDivElement> = useRef(null);
  const [trackStatesFromStorage, setTrackStatesFromStorage] = useState<
    TrackStates
  >({});

  const dispatchBrowserLocation = (chrLocation: ChrLocation) => {
    if (browserRef.current) {
      props.changeBrowserLocation(chrLocation, browserRef.current);
    }
  };

  const changeSelectedSpecies = (genomeId: string) => {
    const chrLocationForGenome = props.chrLocation[genomeId];

    let newUrl: string;
    if (chrLocationForGenome) {
      const params = {
        genomeId,
        focus: props.activeEnsObjectId[genomeId],
        location: getChrLocationStr(chrLocationForGenome)
      };
      newUrl = urlFor.browser(params);
    } else {
      newUrl = urlFor.browser({ genomeId });
      props.fetchExampleEnsObjects(genomeId);
    }
    props.updateBrowserActiveGenomeIdAndSave(genomeId);
    props.replace(newUrl);
  };

  useEffect(() => {
    if (!props.match.params.genomeId) {
      const { activeGenomeId, committedSpecies } = props;
      if (
        find(
          committedSpecies,
          ({ genome_id }: CommittedItem) => genome_id === activeGenomeId
        )
      ) {
        changeSelectedSpecies(activeGenomeId);
      } else {
        changeSelectedSpecies(props.committedSpecies[0].genome_id);
      }
    }
    setTrackStatesFromStorage(browserStorageService.getTrackStates());
  }, [props.match.params.genomeId]);

  useEffect(() => {
    const { genomeId } = props.match.params;

    if (!genomeId) {
      return;
    }

    props.updateBrowserActiveGenomeIdAndSave(genomeId);

    props.fetchGenomeTrackCategories(genomeId);
    props.fetchGenomeInfo(genomeId);
  }, [props.match.params.genomeId]);

  useEffect(() => {
    props.fetchExampleEnsObjects(props.activeGenomeId);
  }, [props.genomeInfo]);

  useEffect(() => {
    const { focus, location } = props.browserQueryParams;
    const { genomeId } = props.match.params;
    const parsedLocation = location && getChrLocationFromStr(location);

    if (!parsedLocation || !focus) {
      return;
    }

    dispatchBrowserLocation(parsedLocation);
    props.fetchEnsObject(focus, genomeId);
    props.fetchEnsObjectTracks(focus, genomeId);
    props.updateBrowserActiveEnsObjectIdAndSave(focus);
  }, [props.browserQueryParams.focus]);

  useEffect(() => {
    const chrLocationForGenome = props.chrLocation[props.activeGenomeId] || [
      '',
      0,
      0
    ];
    const [, chrStart, chrEnd] = chrLocationForGenome;

    if (props.browserActivated && chrStart > 0 && chrEnd > 0) {
      dispatchBrowserLocation(chrLocationForGenome);
    }
  }, [props.browserActivated]);

  const updateLocationInUrl = () => {
    const {
      match: {
        params: { genomeId }
      }
    } = props;
    if (!genomeId) {
      // there is no sense in updating location parameters in url
      // if we donâ€™t yet know genome id
      return;
    }

    let { focus } = props.browserQueryParams;
    if (!focus && props.activeEnsObjectId[genomeId]) {
      focus = props.activeEnsObjectId[genomeId];
    }
    let chrLocationForGenome = props.chrLocation[genomeId];

    if (!chrLocationForGenome && props.browserQueryParams.location) {
      chrLocationForGenome = getChrLocationFromStr(
        props.browserQueryParams.location
      );
    }
    const location = getChrLocationStr(chrLocationForGenome);

    const newUrl = urlFor.browser({ genomeId, focus, location });
    props.replace(newUrl);
  };

  useEffect(() => {
    updateLocationInUrl();
  }, [props.chrLocation, props.browserQueryParams.location]);

  const closeTrack = () => {
    if (props.drawerOpened === false) {
      return;
    }
    props.toggleDrawer(false);
  };
  const getExampleObjectLinks = () => {
    return Object.values(props.exampleEnsObjects[props.activeGenomeId]).map(
      (exampleObject: EnsObject) => {
        const location = `${exampleObject.location.chromosome}:${exampleObject.location.start}-${exampleObject.location.end}`;
        const path = urlFor.browser({
          genomeId: props.activeGenomeId,
          focus: exampleObject.ensembl_object_id,
          location
        });

        return (
          <dd key={exampleObject.ensembl_object_id}>
            <Link to={path}>
              <span className={styles.objectType}>
                {upperFirst(exampleObject.object_type)}
              </span>
              <span className={styles.objectLabel}>{exampleObject.label}</span>
            </Link>
          </dd>
        );
      }
    );
  };

  const [trackAnimation, setTrackAnimation] = useSpring(() => ({
    config: { tension: 280, friction: 45 },
    height: '100%',
    width: 'calc(-36px + 100vw )'
  }));

  const getBrowserWidth = (): string => {
    if (props.drawerOpened) {
      return 'calc(41px + 0vw)';
    }
    return props.trackPanelOpened
      ? 'calc(-356px + 100vw)'
      : 'calc(-36px + 100vw)';
  };

  useEffect(() => {
    setTrackAnimation({
      width: getBrowserWidth()
    });
  }, [props.drawerOpened, props.trackPanelOpened]);

  const getHeightClass = (launchbarExpanded: boolean): string => {
    return launchbarExpanded ? styles.shorter : styles.taller;
  };

  return (
    <>
      <AppBar
        currentAppName={AppName.GENOME_BROWSER}
        activeGenomeId={props.activeGenomeId}
        onTabSelect={changeSelectedSpecies}
      />

      {props.browserQueryParams.focus === undefined && (
        <section className={styles.browser}>
          <BrowserBar dispatchBrowserLocation={dispatchBrowserLocation} />
          {props.exampleEnsObjects[props.activeGenomeId] ? (
            <dl className={styles.exampleLinks}>{getExampleObjectLinks()}</dl>
          ) : null}
        </section>
      )}
      {props.browserQueryParams.focus !== undefined && (
        <section className={styles.browser}>
          <BrowserBar dispatchBrowserLocation={dispatchBrowserLocation} />
          {props.genomeSelectorActive && (
            <div className={styles.browserOverlay} />
          )}
          <div
            className={`${styles.browserInnerWrapper} ${getHeightClass(
              props.launchbarExpanded
            )}`}
          >
            <animated.div style={trackAnimation}>
              <div className={styles.browserImageWrapper} onClick={closeTrack}>
                {props.browserNavOpened &&
                !props.drawerOpened &&
                browserRef.current ? (
                  <BrowserNavBar browserElement={browserRef.current} />
                ) : null}
                <BrowserImage
                  browserRef={browserRef}
                  trackStates={trackStatesFromStorage}
                />
              </div>
            </animated.div>
            <TrackPanel
              browserRef={browserRef}
              trackStates={trackStatesFromStorage}
            />
          </div>
        </section>
      )}
    </>
  );
};

const mapStateToProps = (state: RootState): StateProps => ({
  activeGenomeId: getBrowserActiveGenomeId(state),
  activeEnsObjectId: getBrowserActiveEnsObjectId(state),
  browserActivated: getBrowserActivated(state),
  browserNavOpened: getBrowserNavOpened(state),
  browserOpenState: getBrowserOpenState(state),
  browserQueryParams: getBrowserQueryParams(state),
  chrLocation: getChrLocation(state),
  drawerOpened: getDrawerOpened(state),
  genomeInfo: getGenomeInfo(state),
  genomeSelectorActive: getGenomeSelectorActive(state),
  trackPanelOpened: getTrackPanelOpened(state),
  launchbarExpanded: getLaunchbarExpanded(state),
  exampleEnsObjects: getExampleEnsObjects(state),
  committedSpecies: getEnabledCommittedSpecies(state)
});

const mapDispatchToProps: DispatchProps = {
  changeBrowserLocation,
  fetchEnsObject,
  fetchEnsObjectTracks,
  fetchExampleEnsObjects,
  fetchGenomeInfo,
  fetchGenomeTrackCategories,
  replace,
  toggleDrawer,
  updateBrowserActiveGenomeIdAndSave,
  updateBrowserNavStates,
  updateBrowserActiveEnsObjectIdAndSave
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Browser)
);
