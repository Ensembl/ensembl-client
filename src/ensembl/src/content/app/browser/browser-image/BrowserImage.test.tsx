/**
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import configureMockStore from 'redux-mock-store';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import set from 'lodash/fp/set';

import { createMockBrowserState } from 'tests/fixtures/browser';

import { BrowserImage } from './BrowserImage';

import MockGenomeBrowser from 'tests/mocks/mockGenomeBrowser';

const mockGenomeBrowser = jest.fn();

jest.mock('src/content/app/browser/hooks/useGenomeBrowser', () => () => ({
  genomeBrowser: mockGenomeBrowser(),
  activateGenomeBrowser: jest.fn()
}));

jest.mock('../browser-cog/BrowserCogList', () => () => (
  <div id="browserCogList" />
));

jest.mock('src/content/app/browser/zmenu', () => ({
  ZmenuController: () => <div id="zmenuController" />
}));

jest.mock('src/shared/components/loader', () => ({
  CircleLoader: () => <div id="circleLoader" />
}));

jest.mock('src/shared/components/overlay/Overlay', () => () => (
  <div id="overlay" />
));

const mockState = createMockBrowserState();

const mockStore = configureMockStore([thunk]);

let store: ReturnType<typeof mockStore>;

const renderComponent = (state: typeof mockState = mockState) => {
  store = mockStore(state);
  return render(
    <Provider store={store}>
      <BrowserImage />
    </Provider>
  );
};

describe('<BrowserImage />', () => {
  afterEach(() => {
    mockGenomeBrowser.mockReturnValue(new MockGenomeBrowser());
    jest.resetAllMocks();
  });

  describe('rendering', () => {
    it('renders loader if browser is not activated', () => {
      mockGenomeBrowser.mockReturnValueOnce(undefined);
      const { container } = renderComponent();
      expect(container.querySelector('#circleLoader')).toBeTruthy();
    });

    it('renders browser cog list', () => {
      const { container } = renderComponent();
      expect(container.querySelector('#browserCogList')).toBeTruthy();
    });

    it('renders zmenu controller', () => {
      const { container } = renderComponent();
      expect(container.querySelector('#zmenuController')).toBeTruthy();
    });

    it('has an overlay on top when disabled', () => {
      const { container } = renderComponent(
        set('browser.browserLocation.regionEditorActive', true, mockState)
      );
      expect(container.querySelector('#overlay')).toBeTruthy();
    });
  });
});
