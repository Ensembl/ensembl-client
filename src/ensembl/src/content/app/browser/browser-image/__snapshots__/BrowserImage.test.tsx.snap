// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrowserImage /> renders correctly 1`] = `
<BrowserImage>
  <div
    className="browser-stage"
  />
</BrowserImage>
`;
